# Ejercicio 13: Cultura.
Una persona tiene un libro de cada uno de los autores siguientes: Mario Vargas Llosa, Alberto Ruy Sánchez, Gabriel García Márquez, Günther Grass, Patrick Süskind, Heinrich Böll, Arturo Pérez-Reverte y Camilo José Cela. Como aún no los ha leído, quiere ponerlos en una repisa cerca de su cama. De acuerdo con esto, determina los siguientes puntos:
       
- A) ¿De cuántas formas puede acomodar los libros sobre la repisa?
- B) ¿De cuántas maneras puede ordenarlos si quiere que los libros de autores latinoamericanos estén juntos?
- C) ¿De cuántas formas puede acomodarlos, si los autores latinoamericanos deben estar juntos, así como los autores alemanes y los españoles también?
- D) Determina las probabilidades correspondientes a los incisos (b) y (c).


----

Este ejercicio nos presenta un escenario de combinatoria y probabilidad. **La combinatoria** se encarga de estudiar el número de formas en que se pueden elegir o disponer elementos de un conjunto. En este caso, queremos saber cuántas formas hay de ordenar los libros en la repisa de acuerdo a diferentes restricciones. Por otro lado, la **probabilidad** nos permite calcular la posibilidad de que ocurra un evento específico. En este caso, queremos calcular la probabilidad de que los libros de autores latinoamericanos, alemanes y españoles se encuentren juntos.

In [97]:
# Importación de la biblioteca math como 'm'
import math as m

# Función para los cálculos
def calcular_cultura():
    # Calcular valores (inciso A, B, C) 
    # Número total de libros
    total_libros = 8
    
    # Calcular el número total de formas de arreglar los libros (Inciso A)
    # Utilizando la función factorial de la biblioteca math
    formas_total = m.factorial(total_libros)
    
    # Calcular el número de formas de arreglar los libros con autores latinoamericanos juntos (Inciso B)
    # Hay 3 autores latinoamericanos y 5 autores no latinoamericanos
    # Calcular el número de formas de arreglar los 5 autores no latinoamericanos y los 3 autores latinoamericanos
    formas_b = m.factorial(6) * m.factorial(3)
    
    # Calcular el número de formas de arreglar los libros con autores latinoamericanos, alemanes y españoles juntos (Inciso C)
    # Hay 3 autores latinoamericanos, 2 autores alemanes y 3 autores españoles
    # Calcular el número de formas de arreglar cada grupo de autores y multiplicarlos entre sí
    formas_c = m.factorial(3) * (m.factorial(2) * m.factorial(3) * m.factorial(3))
    
    # Calcular las probabilidades (Inciso D)
    # Probabilidad de arreglar los libros con autores latinoamericanos juntos
    probabilidad_b = formas_b / formas_total
    
    # Probabilidad de arreglar los libros con autores latinoamericanos, alemanes y españoles juntos
    probabilidad_c = formas_c / formas_total
    
    # Devolver los valores calculados
    return formas_total, formas_b, formas_c, probabilidad_b, probabilidad_c

# Variables de guardado
a, b, c, d1, d2 = calcular_cultura()

# Imprimir resultados
print("A) Número de formas en las que puedes acomodar los libros sobre la repisa:", a)
print("B) Maneras en las que puedes ordenar los libros si se quiere que libros de autores latinoamericanos estén juntos:", b)
print("C) Formar en las que puedas acomodarlos si los autores latinos están juntos, así como los alemanes y los españoles:", c)
print("D) Probabilidades de incisos B y C: ") 
print("Probabilidad correspondiente al inciso B:", d1)
print("Probabilidad correspondiente al inciso C:", d2)


A) Número de formas en las que puedes acomodar los libros sobre la repisa: 40320
B) Maneras en las que puedes ordenar los libros si se quiere que libros de autores latinoamericanos estén juntos: 4320
C) Formar en las que puedas acomodarlos si los autores latinos están juntos, así como los alemanes y los españoles: 432
D) Probabilidades de incisos B y C: 
Probabilidad correspondiente al inciso B: 0.10714285714285714
Probabilidad correspondiente al inciso C: 0.010714285714285714


- ----- -


### A) ¿De cuántas formas puede acomodar los libros sobre la repisa? 

Para resolver este inciso, podemos utilizar el concepto de **permutación simple**. La permutación simple se refiere al número de formas en que se pueden ordenar un conjunto de elementos **sin importar la repetición**. En este caso, tenemos 8 libros diferentes, por lo que el número total de arreglos posibles se calcula utilizando la fórmula de permutación simple.
$$
Formas \ totales = n!
$$
Donde '**n**' es el número total de elementos (en este caso, 8 libros) y '**!**' representa el factorial de n.
$$
Formas \ totales = 8! = 40,320
$$

In [102]:
print("Número de formas en las que puedes acomodar los libros sobre la repisa:", a)

Número de formas en las que puedes acomodar los libros sobre la repisa: 40320


- ----- -


#### B) ¿De cuántas maneras puede ordenarlos si quiere que los libros de autores latinoamericanos estén juntos?     
   

En este caso, queremos considerar una restricción: los 3 libros de autores latinoamericanos (correspondientes a los autores Mario Vargas Llosa, Alberto Ruy Sánchez y Gabriel García Márquez) deben estar juntos. Para ello, podemos dividir el problema en dos pasos:

Agrupar los libros latinoamericanos: Imaginemos que los 3 libros latinoamericanos se convierten en un solo "**bloque**". De esta manera, tenemos 6 elementos para ordenar: el bloque de libros latinoamericanos y los 3 libros restantes.

Ordenar los bloques y libros restantes: El número de formas de ordenar estos 6 elementos es simplemente una permutación simple de 6 elementos:
$$
Bloques \ y \ libros \ restantes : 6! = 720
$$
Sin embargo, debemos considerar que dentro del bloque de libros latinoamericanos, los 3 libros se pueden ordenar de diferentes maneras. El número de formas de ordenar estos 3 libros es también una permutación simple:
$$
Formas \ dentro \ del \ bloque : 3! = 6
$$
Ahora, el número total de formas de ordenar los libros con los autores latinoamericanos juntos se obtiene multiplicando las formas de ordenar el bloque y los libros restantes por las formas de ordenar dentro del bloque:

Formas con autores latinoamericanos juntos:     
$$
 Bloques \ y \ libros \ restantes * Formas \ dentro \ del \ bloque: 720 * 6 = 4320
$$

In [103]:
print("Maneras en las que puedes ordenar los libros si se quiere que libros de autores latinoamericanos estén juntos:", b)

Maneras en las que puedes ordenar los libros si se quiere que libros de autores latinoamericanos estén juntos: 4320


- ----- -


#### C) ¿De cuántas formas puede acomodarlos, si los autores latinoamericanos deben estar juntos, así como los autores alemanes y los españoles también?      
  

En este caso, tenemos tres restricciones:

1. Los 3 libros latinoamericanos deben estar juntos (Mario Vargas Llosa, Alberto Ruy Sánchez y Gabriel García Márquez).
2. Los 3 libros alemanes deben estar juntos (Günther Grass, Heinrich Böll y Patrick Süskind).
3. Los 2 libros españoles deben estar juntos (Arturo Pérez Reverte y Camilo José Cela).
     
Para resolver este problema, podemos seguir una estrategia similar al inciso B, pero agrupando los libros por nacionalidad.

Agrupar los libros por nacionalidad: Se forman 3 bloques: bloque de autores latinoamericanos, bloque de autores alemanes y bloque de autores españoles.

Ordenar los bloques: El número de formas de ordenar estos 3 bloques es una permutación simple de 3 elementos:
$$
Formas \ de \ bloques : 3! = 6
$$
Ordenar dentro de cada bloque: Se calcula el número de formas de ordenar los libros dentro de cada bloque utilizando la permutación simple:

- Formas dentro del bloque latinoamericano: $$3! = 6$$        
- Formas dentro del bloque alemán: $$3! = 6$$     
- Formas dentro del bloque español: $$2! = 2$$        
- Multiplicar las formas: El número total de formas de ordenar los libros con las restricciones dadas se obtiene multiplicando las formas de ordenar los bloques por las formas de ordenar dentro de cada bloque:

Formas con restricciones:
$$
Formas \ de \ bloques * Formas \ bloque \ latinoamericano * Formas \ bloque \ alemán * Formas \ bloque \ español : 6 * 6 * 6 * 2 = 432
$$

In [104]:
print("Formar en las que puedas acomodarlos si los autores latinos están juntos, así como los alemanes y los españoles:", c)

Formar en las que puedas acomodarlos si los autores latinos están juntos, así como los alemanes y los españoles: 432


- ----- -


#### D) Determina las probabilidades correspondientes a los incisos (b) y (c).  

En este inciso, queremos calcular la probabilidad de que se cumplan las restricciones de agrupación de libros especificadas en los incisos B y C. Para ello, utilizaremos el concepto de **probabilidad**, que se define como la proporción de eventos favorables con respecto al espacio muestral total.

- Espacio muestral:

El espacio muestral, en este caso, es el conjunto de todas las formas posibles de ordenar los 8 libros sin ninguna restricción. Ya hemos calculado en el inciso A que el número total de formas posibles es de 40.320, utilizando la fórmula de permutación simple:
$$
Formas \ totales : 8! = 40320
$$
- Eventos favorables:

    - **Inciso B:** El evento favorable es que los 4 libros de autores latinoamericanos se encuentren juntos. Ya hemos calculado en el inciso B que el número de formas de ordenar los libros con esta restricción es de 4320.
    $$
    Bloques \ y \ libros \ restantes : 6! = 720
    $$
    $$
    Formas \ dentro \ del \ bloque : 3! = 6
    $$
    $$
    Bloques \ y \ libros \ restantes * Formas \ dentro \ del \ bloque: 720 * 6 = 4320
    $$

    - **Inciso C:** El evento favorable es que los grupos de autores latinoamericanos, alemanes y españoles se encuentren juntos. Ya hemos calculado en el inciso C que el número de formas de ordenar los libros con esta restricción es de 432.

        - Formas dentro del bloque latinoamericano: $$3! = 6$$        
        - Formas dentro del bloque alemán: $$3! = 6$$     
        - Formas dentro del bloque español: $$2! = 2$$        
        - Formas con restricciones:
        $$
        Formas \ de \ bloques * Formas \ bloque \ latinoamericano * Formas \ bloque \ alemán * Formas \ bloque \ español : 6 * 6 * 6 * 2 = 432
        $$

- Cálculo de probabilidades:

    - **Inciso B:**  
        $$
        P(\text{Autores Latinoamericanos juntos}) = \frac{\text{Formas con autores latinoamericanos juntos}}{\text{Formas totales}} = \frac{4320}{40320} = \frac{1}{9} \approx 0.10714285714285714
        $$
    - **Inciso C:**
        $$
        P(\text{Grupos Juntos}) = \frac{\text{Número de maneras con latinoamericanos, alemanes y españoles juntos}}{\text{{Formas totales}}} = \frac{432}{40320} = \frac{1}{93.333} \approx 0.010714285714285714
        $$

        
B) Existe una probabilidad del 0.10714285714285714 % de que los libros de autores latinoamericanos se encuentren juntos en la repisa si el lector no impone ninguna otra restricción sobre el orden de los libros.

C) Existe una probabilidad del 0.010714285714285714 % de que los grupos de autores latinoamericanos, alemanes y españoles se encuentren juntos en la repisa si el lector impone la restricción de que los grupos de autores deben estar juntos, pero no la restricción de que los autores dentro de cada grupo deban estar juntos.


In [105]:
print("Probabilidad correspondiente al inciso B:", d1)
print("Probabilidad correspondiente al inciso C:", d2)

Probabilidad correspondiente al inciso B: 0.10714285714285714
Probabilidad correspondiente al inciso C: 0.010714285714285714
